src/group_theory/noncomm_pi_coprod.lean

```diff
@@ -0,0 +1,426 @@
+    ... = ϕ j (f j) * (ϕ i (g i) * to_fun ϕ hcomm f S) : by rw mul_assoc
+    ... = ϕ j (f j) * (to_fun ϕ hcomm f S * ϕ i (g i)) : by { congr' 1, apply ih hiS }
+    ... = (ϕ j (f j) * to_fun ϕ hcomm f S) * ϕ i (g i) : by rw ← mul_assoc }
+end
+
+@[simp, to_additive add_to_fun_add]
+lemma to_fun_mul (S : finset ι) :
+  to_fun ϕ hcomm (f * g) S = to_fun ϕ hcomm f S * to_fun ϕ hcomm g S :=
+begin
+  classical,
+  induction S using finset.induction_on with i S hnmem ih,
+  { simp, },
+  { simp only [to_fun_insert_of_not_mem _ _ _ _ hnmem],
+    rw ih, clear ih,
+    simp only [pi.mul_apply, map_mul],
+    repeat { rw mul_assoc }, congr' 1,
+    repeat { rw ← mul_assoc }, congr' 1,
+    exact (to_fun_commutes _ _ _ S i hnmem), }
+end
+
+@[to_additive add_to_fun_in_sup_mrange]
+lemma to_fun_in_sup_mrange (S : finset ι) :
+  to_fun ϕ hcomm f S ∈ ⨆ i ∈ S, (ϕ i).mrange :=
+begin
+  classical,
+  induction S using finset.induction_on with i S hnmem ih,
+  { simp, },
+  { simp only [to_fun_insert_of_not_mem _ _ _ _ hnmem],
+    refine (submonoid.mul_mem _ _ _),
+    { apply submonoid.mem_Sup_of_mem, { use i }, { simp, }, },
+    { refine @bsupr_le_bsupr' _ _ _ _ _ _ (λ i, (ϕ i).mrange) _ ih,
+      exact λ i, finset.mem_insert_of_mem } }
+end
+
+variable (hcomm)
+
+/-- The canonical homomorphism from a family of monoids, restricted to a subset of the index space.
+-/
+@[to_additive add_hom "The canonical homomorphism from a family of additive monoids, restricted to a
+subset of the index space"]
+def hom : (Π (i : ι), N i) →* M :=
+{ to_fun := λ f, to_fun ϕ hcomm f S,
+  map_one' := to_fun_one ϕ _,
+  map_mul' := λ f g, to_fun_mul ϕ f g S, }
+
+variable {hcomm}
+
+include hdec
+
+@[to_additive add_to_fun_single]
+lemma to_fun_single (i : ι) (y : N i) (S : finset ι) :
+  to_fun ϕ hcomm (monoid_hom.single _ i y) S = if i ∈ S then ϕ i y else 1 :=
+begin
+  induction S using finset.induction_on with j S hnmem ih,
+  { simp, },
+  { simp only [ to_fun_insert_of_not_mem _ _ _ _ hnmem],
+    by_cases (i = j),
+    { subst h, rw ih, simp [hnmem], },
+    { change i ≠ j at h,
+      simpa [h] using ih, } }
+end
+
+@[simp, to_additive add_hom_single]
+lemma hom_single (i : ι) (y : N i):
+  hom ϕ hcomm S (monoid_hom.single _ i y) = if i ∈ S then ϕ i y else 1 := to_fun_single _ _ _ _
+
+omit hdec
+
+@[to_additive add_mrange]
+lemma mrange : (hom ϕ hcomm S).mrange = ⨆ i ∈ S, (ϕ i).mrange :=
+begin
+  classical,
+  apply le_antisymm,
+  { rintro x ⟨f, rfl⟩,
+    exact (to_fun_in_sup_mrange ϕ f S), },
+  { refine (bsupr_le _),
+    rintro i hmem x ⟨y, rfl⟩,
+    use (monoid_hom.single _ i y),
+    simp [hmem], }
+end
+
+end noncomm_pi_coprod_on
+
+namespace monoid_hom
+
+include hfin
+
+variable (hcomm)
+
+/-- The canonical homomorphism from a family of monoids. -/
+@[to_additive "The canonical homomorphism from a family of additive monoids."]
+def noncomm_pi_coprod : (Π (i : ι), N i) →* M := noncomm_pi_coprod_on.hom ϕ hcomm finset.univ
```

Comment:
To match `linear_map.lsum`, is this true as an equiv?

```lean
def noncomm_pi_coprod_equiv :
  {ϕ : Π i, N i →* M // ∀ (i j : ι), i ≠ j → ∀ (x : N i) (y : N j), commute (ϕ i x) (ϕ j y) }
    ≃ ((Π i, N i) →* M) :=
```

Reply:
Upon first glance, I think it is. I'll give it a shot!

Reply:
I can show it if I have
```
@[to_additive]
lemma _root_.monoid_hom.pi_ext [decidable_eq ι] [fintype ι] {f g : (Π i, N i) →* M} :
  f = g ↔ (∀ i x, f (monoid_hom.single N i x) = g (monoid_hom.single N i x)) :=
sorry
```
which I yet have to prove, and find a home for, or find existing somewhere.